src/sentry/scripts/similarity/index.lua

```diff
@@ -0,0 +1,618 @@
+                            end,
+                            {}
+                        )
+                    end
+                )
+            end
+
+            local fetch_candidates = function (index, frequencies)
+                local candidates = {}  -- acts as a set
+                for band, buckets in ipairs(frequencies) do
+                    for bucket, count in pairs(buckets) do
+                        for _, time in ipairs(time_series) do
+                            -- Fetch all other items that have been
+                            -- added to the same bucket in this bad
+                            -- during this time period.
+                            local members = redis.call(
+                                'SMEMBERS',
+                                get_bucket_membership_key(
+                                    configuration.scope,
+                                    index,
+                                    time,
+                                    band,
+                                    bucket
+                                )
+                            )
+                            for _, member in ipairs(members) do
+                                candidates[member] = true
+                            end
+                        end
+                    end
+                end
+                return candidates
+            end
+
+            return table.imap(
+                indices,
+                function (index)
+                    local item_frequencies = fetch_bucket_frequencies(index, item_key)
+
+                    local candidates = fetch_candidates(index, item_frequencies)
+                    local candidate_frequencies = {}
+                    for candidate_key, _ in pairs(candidates) do
+                        candidate_frequencies[candidate_key] = fetch_bucket_frequencies(
+                            index,
+                            candidate_key
+                        )
+                    end
+
+                    local results = {}
+                    for key, value in pairs(candidate_frequencies) do
+                        table.insert(
+                            results,
+                            {
+                                key,
+                                table.ireduce(
+                                    table.imap(
+                                        table.izip(
+                                            item_frequencies,
+                                            value
+                                        ),
+                                        function (v)
+                                            local dist = get_manhattan_distance(
+                                                scale_to_total(v[1]),
+                                                scale_to_total(v[2])
+                                            )
+                                            return 1 - (dist / 2)
```

Comment:
FWIW, I chose not to do this (for now at last) because I think it makes it harder to understand what these constants are for (and I'm not too worried about the overhead of the mathematical operations here.) This is clever though!